state/optimistic_syncing/pipeline.go

```diff
@@ -0,0 +1,474 @@
+	ParentState State
+}
+
+// Config contains the configuration for the pipeline
+type Config struct {
+	// Logger is the logger to use for the pipeline
+	Logger zerolog.Logger
+	// IsSealed indicates if the data is sealed
+	IsSealed bool
+	// ExecutionResultID is a unique identifier for the execution result being processed
+	ExecutionResultID string
+	// BlockID is the identifier of the block being processed
+	BlockID string
+	// Core implements the processing logic for the pipeline
+	Core Core
+	// ChildrenStateUpdateChans are channels to send state updates to children
+	ChildrenStateUpdateChans []chan<- StateUpdate
+}
+
+// Pipeline represents a generic processing pipeline with state transitions.
+// It processes data through sequential states: Ready -> Downloading -> Indexing ->
+// WaitingPersist -> Persisting -> Complete, with conditions for each transition.
+type Pipeline struct {
+	logger              zerolog.Logger
+	stateUpdateChan     chan StateUpdate
+	childrenUpdateChans []chan<- StateUpdate
+
+	mu                 sync.RWMutex
+	state              State
+	isSealed           bool
+	descendsFromSealed bool
+	parentState        State
+	executionResultID  string
+	blockID            string
+
+	core          Core
+	stateNotifier engine.Notifier
+}
+
+// NewPipeline creates a new processing pipeline.
+// The pipeline is always initialized in the Ready state and will begin processing
+// when Run is called.
+func NewPipeline(config Config) *Pipeline {
+	p := &Pipeline{
+		logger:              config.Logger.With().Str("component", "pipeline").Str("executionResultID", config.ExecutionResultID).Logger(),
+		stateUpdateChan:     make(chan StateUpdate, 10),
+		childrenUpdateChans: config.ChildrenStateUpdateChans,
+		state:               StateReady,
+		isSealed:            config.IsSealed,
+		descendsFromSealed:  true,
+		stateNotifier:       engine.NewNotifier(),
+		core:                config.Core,
+		executionResultID:   config.ExecutionResultID,
+		blockID:             config.BlockID,
+	}
+
+	return p
+}
+
+// Run starts the pipeline processing and blocks until completion or context cancellation.
+//
+// This function handles the progression through the pipeline states, executing the appropriate
+// processing functions at each step. It also listens for state updates from the parent pipeline
+// and cancels processing if the pipeline no longer descends from the latest persisted result.
+//
+// When the pipeline reaches a terminal state (StateComplete or StateCanceled), the function returns.
+// The function will also return if the provided context is canceled.
+//
+// Returns an error if any processing step fails or if the context is canceled.
+func (p *Pipeline) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go p.listenForStateUpdates(ctx, cancel)
+
+	notifierChan := p.stateNotifier.Channel()
+
+	// Trigger initial check
+	p.stateNotifier.Notify()
+
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
this needs to use the context passed to the function, not the context returned from `context.WithCancel`. otherwise, this loop will abort when a state update is received making the pipeline abandoned.